### ISSUE TITLE
Prevent resource:// Leak - Fingerprinting

### ISSUE THREAD

**Issue by ilikenwf** [MoonchildProductions/Pale-Moon#445]
This particular bug allows 3rd parties to get browser details including installed addons. A friendly dev has created a rough version for Pale Moon, but also suggested a way we can go about adding it into the core to fix the bug outright.

- **Test**: https://www.browserleaks.com/firefox
- **PM Addon (test, but works):**  https://notabug.org/neuemars/xpi-bucket/raw/master/palemoon/%40no-resource-uri-leak-0.2.1-pm.xpi

Quote from the addon author:

> ...this problem needs a browser-level fix. Let's spread words about
> it so that more Mozilla-related browser developers become aware of it.
> Pale Moon devs may want to modify ScriptSecurityManager or something to
> fix the root of the problem. For reference, this is the Bugzilla bug at
> Mozilla:
> https://bugzilla.mozilla.org/show_bug.cgi?id=863246

**Comment by wolfbeast** [comment]
I'm keeping an eye on the relevant BZ bug. The extension is a good stop-gap measure until something is implemented in the browser core.

**Comment by ilikenwf** [comment]
Sounds good.

The only issue I noticed was that for whatever reason, at least on my
local/private gitweb setup (which uses a custom template but worked fine
before installing the addon, and fine if removed), this happens when
visiting the repo pages:

https://i.imgur.com/1r1y3vQ.png

**Comment by wolfbeast** [comment]
That would be something to discuss with the add-on dev :)

**Comment by ilikenwf** [comment]
Oh trust me, I am. I'm really happy he noticed and worked around this
really old bug.

**Comment by beerisgood** [comment]
Would be awesome if his addon works nativ with Pale Moon.
Here the issue entry on their website: https://notabug.org/desktopd/no-resource-uri-leak/issues/2

> However, Pale Moon currently doesn't seem to support ECMAScript 2015 Edition (ES6). Also, Pale Moon should include (without supporting Australis part) support for JPM-based add-ons. Many incompatible add-ons are just so because Pale Moon does not support JPM. (Just a mater of adding some CommonJS modules. No need to support Australis.)

**Comment by LimboSlam** [comment]
Hm, it seems ES6 is becoming more and more needed among add-on developers. 

Question @wolfbeast: Why doesn't Pale Moon support these JPM-based add-ons and will this be included in the re-base?

**Comment by wolfbeast** [comment]
Not a single browser out there supports ES6 in its entirety so far. It's massive, and it's also nonsense to implement all of it. In addition, pretty much everything ES6 offers can be implemented in JS itself with a little effort, even OOP things like Promises.

Additionally, and people really need to get this through their skull, **it is not Pale Moon's task to make itself compatible with what is supposed to extend it.** If you write an extension, then you aim to _extend_ a specific browser. That is your _target_ application. If that is Firefox with all its specific features, then that's fine, but don't expect us to become "more like Firefox" just because an _extension_ needs something that happens to be implemented and used in that other browser.

> The jpm tool is the replacement for cfx. It enables you to test, run, and package add-ons.

It's a different way of packaging add-ons that can't be "magically" added to Pale Moon by dropping in a few libraries. Also, I don't really see a reason why we should support this because all add-ons written and packaged with this tool will be Australis-targeted and likely never be compatible with us.

We are likely not going to support this type of add-on at all in the future because it relies on an abstraction layer that won't be compatible with us any longer - the only way those add-ons can be supported is if a compatible set of SDK libraries is going to be included in the extension itself to make it self-contained and only talks to the browser with native APIs as a result.

Bottom line is: this extension is a temporary measure, has been made compatible with us, and as such is "good enough". I don't expect extended use of this.

**Comment by ilikenwf** [comment]
Beyond breaking our resource:// pages, for "page not found" etc, this add
on does fill the gap. I'm not sure how to fix it myself or I would, I've
been out of the "tinkering with Mozilla and friends" game for about a year,
as I've not touched the Nightingale code in some time.

**Comment by beerisgood** [comment]
The Pale Moon version on AMO (https://addons.mozilla.org/de/firefox/addon/no-resource-uri-leak/) works! (https://outgoing.mozilla.org/v1/cf9180eb148c9338dfaae1043e3c2a7903f5b3c147106d15f918c1af95c1fbef/https%3A//notabug.org/neuemars/xpi-bucket/raw/master/palemoon/%40no-resource-uri-leak-0.2.1-pm.xpi)

Thanks for that work!

**Comment by TPS** [comment]
Just FYI: I pointed out this discussion to the addon author on the official support page.  Ideally, he'll chime in here, especially re: https://github.com/MoonchildProductions/Pale-Moon/issues/445#issuecomment-226959128.

**Comment by mattatobin** [comment]
I would like to point out that codename "Tycho" has dropped support for the Add-ons SDK so this extension as written will not function in the next Milestone version of Pale Moon.

**Comment by beerisgood** [comment]
Then i hope the addon dev give Pale Moon support even with the next Milestone of Pale Moon.

**Comment by LimboSlam** [comment]
Funny @mattatobin, I thought with the re-base that it'll solve this issue, or are you guys merely dropping this type of format? Well it looks like I have a lot to learn still.

**Comment by wolfbeast** [comment]
It's simply not possible to offer an add-on SDK for the re-based browser because it's a hybrid. The existing Mozilla SDKs won't match our code (neither the current or the one included in the re-base) and it's simply not feasible to (1) rewrite the SDK for the resulting browser and (2) expect extension developers to target that SDK (which would be be significantly different than any Firefox version). So we're dropping support for it.

**Comment by wolfbeast** [comment]
Back to the original "problem" with this access: I've looked at the potential attack surface this gives and really, no user data is ever exposed through `resource://` URIs. The only data that can be garnered from inspecting `resource://` URIs is application-static data, none of which is a privacy concern. It would be an extremely expensive way to fingerprint "You're using version X of web browser Y" and the only thing from a tracker point of view it works around is UA-randomizers/spoofers.

Marking this wontfix; there simply isn't enough reason to create an elaborate block+whitelist as is done in the Tor BMO bug to mitigate something that is likely never going to be used at scale in the wild because of the expensive nature of it.
